fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/query/plan/synthetic/SyntheticRecordPlannerTest.java

```diff
@@ -0,0 +1,699 @@
+
+            context.commit();
+        }
+
+        try (FDBRecordContext context = openContext()) {
+            final FDBRecordStore recordStore = recordStoreBuilder.setContext(context).open();
+            final SyntheticRecordPlanner planner = new SyntheticRecordPlanner(recordStore);
+
+            SyntheticRecordPlan plan1 = planner.scanForType(recordStore.getRecordMetaData().getSyntheticRecordType("ManyToMany"));
+            Multiset<Tuple> expected1 = ImmutableMultiset.of(Tuple.from(-1, 1, 1000, 100), Tuple.from(-1, 2, 1000, 101), Tuple.from(-1, 2, 1002, 102));
+            Multiset<Tuple> results1 = HashMultiset.create(plan1.execute(recordStore).map(FDBSyntheticRecord::getPrimaryKey).asList().join());
+            assertEquals(expected1, results1);
+        }
+    }
+
+    @Test
+    public void selfJoin() throws Exception {
+        metaDataBuilder.addIndex("MySimpleRecord", "other_rec_no");
+        final JoinedRecordTypeBuilder joined = metaDataBuilder.addJoinedRecordType("SelfJoin");
+        joined.addConstituent("simple1", "MySimpleRecord");
+        joined.addConstituent("simple2", "MySimpleRecord");
+        joined.addJoin("simple1", "other_rec_no", "simple2", "rec_no");
+
+        try (FDBRecordContext context = openContext()) {
+            final FDBRecordStore recordStore = recordStoreBuilder.setContext(context).create();
+
+            for (int i = 0; i < 3; i++) {
+                TestRecordsJoinIndexProto.MySimpleRecord.Builder simple = TestRecordsJoinIndexProto.MySimpleRecord.newBuilder();
+                simple.setRecNo(i).setOtherRecNo(i + 1);
+                recordStore.saveRecord(simple.build());
+            }
+
+            context.commit();
+        }
+
+        try (FDBRecordContext context = openContext()) {
+            final FDBRecordStore recordStore = recordStoreBuilder.setContext(context).open();
+            final SyntheticRecordPlanner planner = new SyntheticRecordPlanner(recordStore);
+
+            SyntheticRecordPlan plan1 = planner.scanForType(recordStore.getRecordMetaData().getSyntheticRecordType("SelfJoin"));
+            Multiset<Tuple> expected1 = ImmutableMultiset.of(Tuple.from(-1, 0, 1), Tuple.from(-1, 1, 2));
+            Multiset<Tuple> results1 = HashMultiset.create(plan1.execute(recordStore).map(FDBSyntheticRecord::getPrimaryKey).asList().join());
+            assertEquals(expected1, results1);
+
+            FDBStoredRecord<Message> record = recordStore.loadRecord(Tuple.from(1));
+            SyntheticRecordFromStoredRecordPlan plan2 = planner.fromStoredType(record.getRecordType(), false);
+            Multiset<Tuple> expected2 = ImmutableMultiset.of(Tuple.from(-1, 0, 1), Tuple.from(-1, 1, 2));
+            Multiset<Tuple> results2 = HashMultiset.create(plan2.execute(recordStore, record).map(FDBSyntheticRecord::getPrimaryKey).asList().join());
+            assertEquals(expected2, results2);
+        }
+    }
+
+    @Test
+    public void outerJoins() throws Exception {
```

Comment:
Maybe a test for outer joins with record types that have multiple columns to check the `n` nulls thing.

Reply:
The repeated nulls thing is gone now.